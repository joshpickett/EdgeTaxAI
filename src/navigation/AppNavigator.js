import React from "react";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import { useSelector } from 'react-redux';
import { Alert } from "react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";

// Auth Screens
import LoginScreen from "../screens/LoginScreen";
import SignupScreen from "../screens/SignupScreen";
import DashboardScreen from "../screens/DashboardScreen";
import MileageTrackingScreen from "../screens/MileageTrackingScreen";
import ReportsScreen from "../screens/ReportsScreen";
import ExpensesScreen from "../screens/ExpensesScreen";
import BankIntegrationScreen from "../screens/BankIntegrationScreen";
import TaxReminderScreen from "../screens/TaxReminderScreen";
import ProfileScreen from "../screens/ProfileScreen";
import GigPlatformScreen from "../screens/GigPlatformScreen";

// Services
import { logoutUser } from "../services/api";

const Stack = createNativeStackNavigator();

const AppNavigator = ({ onReady }) => {
  const { isAuthenticated } = useSelector(state => state.auth);

  // Logout Logic
  const handleLogout = async (navigation) => {
    try {
      await logoutUser(); // Clears user session and tokens
      await AsyncStorage.removeItem("userToken"); // Remove token locally
      Alert.alert("Success", "You have been logged out.");
      navigation.reset({
        index: 0,
        routes: [{ name: "Login" }],
      });
    } catch (error) {
      Alert.alert("Error", "Failed to log out. Please try again.");
    }
  };

  return (
    <Stack.Navigator>
      {!isAuthenticated ? (
        <Stack.Group screenOptions={{ headerShown: false }}>
          <Stack.Screen name="Login" component={LoginScreen} />
          <Stack.Screen name="Signup" component={SignupScreen} />
        </Stack.Group>
      ) : (
        <Stack.Group>
          <Stack.Screen
            name="Dashboard"
            component={DashboardScreen}
            options={{
              title: "Dashboard",
              headerRight: ({ navigation }) => (
                <LogoutButton onPress={() => handleLogout(navigation)} />
              ),
            }}
          />
          <Stack.Screen
            name="MileageTracking"
            component={MileageTrackingScreen}
            options={{ title: "Mileage Tracking" }}
          />
          <Stack.Screen
            name="Expenses"
            component={ExpensesScreen}
            options={{ title: "Manage Expenses" }}
          />
          <Stack.Screen
            name="Reports"
            component={ReportsScreen}
            options={{ title: "Reports" }}
          />
          <Stack.Screen
            name="BankIntegration"
            component={BankIntegrationScreen}
            options={{ title: "Bank Integration" }}
          />
          <Stack.Screen
            name="TaxReminder"
            component={TaxReminderScreen}
            options={{ title: "Tax Reminders" }}
          />
          <Stack.Screen
            name="Profile"
            component={ProfileScreen}
            options={{ title: "User Profile" }}
          />
          <Stack.Screen
            name="GigPlatforms"
            component={GigPlatformScreen}
            options={{ title: "Manage Gig Platforms" }}
          />
        </Stack.Group>
      )}
    </Stack.Navigator>
  );
};

export default AppNavigator;
