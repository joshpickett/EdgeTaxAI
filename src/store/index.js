import { configureStore } from '@reduxjs/toolkit';
import authReducer from './slices/authSlice';
import expenseReducer from './slices/expenseSlice';
import taxReducer from './slices/taxSlice';
import bankIntegrationReducer from './slices/bankIntegrationSlice';
import gigPlatformReducer from './slices/gigPlatformSlice';

export const store = configureStore({
  reducer: {
    auth: authReducer,
    expenses: expenseReducer,
    tax: taxReducer,
    bankIntegration: bankIntegrationReducer,
    gigPlatform: gigPlatformReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
