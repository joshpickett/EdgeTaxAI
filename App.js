import React, { useEffect } from "react";
import { View, Text, StyleSheet, Alert } from "react-native";
import firebase from "firebase/app";
import "firebase/messaging";
import * as Notifications from "expo-notifications";
import Constants from "expo-constants";
import AppNavigator from "./src/navigation/AppNavigator";
import { registerForPushNotifications } from "./src/services/notificationService"; // Push Notification Logic

// Load environment variables
import {
  FIREBASE_API_KEY,
  FIREBASE_AUTH_DOMAIN,
  FIREBASE_PROJECT_ID,
  FIREBASE_STORAGE_BUCKET,
  FIREBASE_MESSAGING_SENDER_ID,
  FIREBASE_APP_ID,
} from "react-native-dotenv";

// Firebase Configuration
const firebaseConfig = {
  apiKey: FIREBASE_API_KEY,
  authDomain: FIREBASE_AUTH_DOMAIN,
  projectId: FIREBASE_PROJECT_ID,
  storageBucket: FIREBASE_STORAGE_BUCKET,
  messagingSenderId: FIREBASE_MESSAGING_SENDER_ID,
  appId: FIREBASE_APP_ID,
};

try {
  if (!firebase.apps.length) {
    firebase.initializeApp(firebaseConfig);
    console.log("Firebase initialized successfully!");
  }
} catch (error) {
  console.error("Firebase initialization error:", error.message);
}

const App = () => {
  useEffect(() => {
    const setupNotifications = async () => {
      try {
        const { expoPushToken, fcmToken } = await registerForPushNotifications();
        console.log("Expo Push Token:", expoPushToken);
        console.log("FCM Token:", fcmToken);
      } catch (error) {
        console.error("Notification Setup Error:", error.message);
      }
    };

    setupNotifications();
  }, []);

  return (
    <View style={styles.container}>
      <AppNavigator />
      <Text style={styles.text}>Welcome to EdgeTaxAI!</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1, justifyContent: "center", alignItems: "center" },
  text: { fontSize: 18, fontWeight: "bold", color: "#333" },
});

export default App;
