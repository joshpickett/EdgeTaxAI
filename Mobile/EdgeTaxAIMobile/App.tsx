import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { Provider, useSelector } from 'react-redux';
import { store } from './src/store';
import AppNavigator from './src/navigation/AppNavigator';
import { colors } from './src/styles/tokens';
import { sharedStyles } from './src/styles/shared';
import {
  SafeAreaView,
  StatusBar,
  useColorScheme,
  View,
} from 'react-native';
import ErrorBoundary from './src/components/ErrorBoundary';
import LoadingOverlay from './src/components/LoadingOverlay';
import { useAuth } from './src/hooks/useAuth';

function App(): React.JSX.Element {
  const isDarkMode = useColorScheme() === 'dark';
  const [isLoading, setIsLoading] = React.useState(true);
  const { initializeAuth } = useAuth();

  React.useEffect(() => {
    const initialize = async () => {
      await initializeAuth();
      setIsLoading(false);
    };
    initialize();
  }, []);

  const backgroundStyle = {
    backgroundColor: isDarkMode ? colors.darkMode.background.default : colors.grey[50],
    flex: 1,
  };

  const barStyle = isDarkMode ? 'light-content' : 'dark-content';

  return (
    <Provider store={store}>
      <NavigationContainer>
        <ErrorBoundary>
          <SafeAreaView style={[backgroundStyle, sharedStyles.safeArea]}>
            <StatusBar
              barStyle={barStyle}
              backgroundColor={backgroundStyle.backgroundColor}
            />
            {isLoading && <LoadingOverlay />}
            <View style={[backgroundStyle, { flex: 1 }]} testID="app-root">
              <AppNavigator />
            </View>
          </SafeAreaView>
        </ErrorBoundary>
      </NavigationContainer>
    </Provider>
  );
}

export default App;
