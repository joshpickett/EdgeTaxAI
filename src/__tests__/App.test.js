import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../store';
import App from '../../App';
import AppNavigator from '../navigation/AppNavigator';

// Mock the AppNavigator component
jest.mock('../navigation/AppNavigator', () => {
  return function MockAppNavigator() {
    return <div data-testid="mock-app-navigator">Mock Navigator</div>;
  };
});

describe('App Component', () => {
  it('renders without crashing', () => {
    const { container } = render(<App />);
    expect(container).toBeTruthy();
  });

  it('renders with Redux Provider', () => {
    const { container } = render(<App />);
    // Check if Provider exists by verifying store context
    expect(container.firstChild).toBeTruthy();
  });

  it('renders AppNavigator component', () => {
    const { getByTestId } = render(<App />);
    const navigator = getByTestId('mock-app-navigator');
    expect(navigator).toBeInTheDocument();
  });

  it('provides Redux store to child components', () => {
    const mockStore = {
      ...store,
      getState: jest.fn(),
      dispatch: jest.fn(),
    };

    const { container } = render(
      <Provider store={mockStore}>
        <AppNavigator />
      </Provider>
    );
    
    expect(container).toBeTruthy();
  });

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  // Error boundary test
  it('handles errors gracefully', () => {
    const spy = jest.spyOn(console, 'error');
    spy.mockImplementation(() => {});

    const ErrorComponent = () => {
      throw new Error('Test error');
    };

    jest.mock('../navigation/AppNavigator', () => ErrorComponent);

    expect(() => render(<App />)).not.toThrow();

    spy.mockRestore();
  });
});
