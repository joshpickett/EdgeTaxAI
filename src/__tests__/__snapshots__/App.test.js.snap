// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-app-navigator"
  >
    Mock Navigator
  </div>
</div>
`;
